File: primes.cpp
#include <stdlib.h>
#include <math.h>

#include <cmath>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <chrono>
#include <algorithm>
#include <locale>
#include <iomanip>

#include <boost/dynamic_bitset.hpp>

typedef uint64_t T;
typedef std::chrono::steady_clock _clock;

auto starttime = _clock::now();

T sqr(T x) {
    return x*x;
};

class PrimeCalculator {

    public:
    std::vector<uint32_t> P = {3,5,7};

    PrimeCalculator() {
        P.reserve(1 << 28);
    };

    /* Compute primes between p[cur]^2 and p[cur+1]^2, excluding. Resizes c so
     * it holds a bool for each odd number in that range. Performs the sieve
     * on this range and returns the offset such that c[i] contains whether
     * offset+2*i is prime or not.
     */
    size_t sieve(size_t cur, boost::dynamic_bitset<> &c) const {
        const T start = sqr(T(P[cur])) + 2;
        const T end = sqr(T(P[cur+1]));
        c.clear();
        c.resize((end-start)/2, true);

        for (size_t i=0; i<=cur; i++) {
            const auto p = T(P[i]);
            T first = p*((start-1)/p+1);  // first multiple of p in range
            // we only check odd numbers
            if (first % 2 == 0)
                first += p;
            for (size_t idx = (first-start)/2; idx < c.size(); idx += p) {
                c[idx] = false;
            }
        };
        return start;
    }

    /* Compute generating primes between 9 and P[endidx]^2 and fill P with
     * them.*/
    void compute(size_t endidx) {
        boost::dynamic_bitset<> c;
        for (size_t cur = 0; cur<endidx; cur++) {
            auto start = sieve(cur, c);
            for (size_t i = 0; i<c.size(); i++) {
                if (c[i])
                    P.push_back(start+2*i);
            }
        }
    }

    /* Count primes between P[startidx]^2 and P[endidx]^2*/
    size_t count(size_t startidx, size_t endidx) const {
        size_t result = 0;
        boost::dynamic_bitset<> c;
        for (size_t cur = startidx; cur<endidx; cur++) {
            sieve(cur, c);
            result += c.count();
        }
        return result;
    }

    void print(size_t startidx, size_t endidx, T result, T end=0) const {
        if (endidx)
            end = P[endidx];

        std::cout
            << std::fixed << std::chrono::duration<double>(_clock::now() - starttime).count() << "s: "
            << "π(" << std::setw(12);
        if (endidx)
            std::cout << P[endidx] << "²";
        else
            std::cout << end;
        std::cout << ") - "
            << "π(" << std::setw(12) << P[startidx] << "²) = "
            << result
            << std::endl;
    }
};

int main(int argc, char** argv) {
    std::cout.imbue(std::locale(""));
    PrimeCalculator p;
    auto &P = p.P;
    // 65521=P[6540] is the largest prime below 2^16
    size_t cur = 6540;
    p.compute(cur);
    p.print(0, cur, P.size()-3);

    #pragma omp parallel
    {
        while (true) {
            size_t local_cur;
            size_t tgt;
            #pragma omp critical
            {
                local_cur = cur;
                tgt = cur + 1024;
                cur = tgt;
            }
            size_t count = p.count(local_cur, tgt);
            #pragma omp critical
            p.print(local_cur, tgt, count);
        }
    }
    return 0;
}

File: primes.rs
use std::time::Instant;
use crossbeam::{channel::bounded,thread::scope};
use bitvec::prelude::*;

fn sqr(x: u32) -> u64 {
    return (x as u64)*(x as u64);
}

struct Calc {
    p: Vec<u32>,
    start_time: Instant,
}

impl Calc {
    fn new() -> Calc {
        let mut vec = Vec::<u32>::with_capacity(1<<28);
        vec.push(3);
        vec.push(5);
        vec.push(7);
        Calc{p: vec, start_time: Instant::now()}
    }
    /* Compute primes between p[cur]^2 and p[cur+1]^2, excluding. Resizes c so it holds a bool for
     * each odd number in that range. Performes the sieve on this range and returns the offset
     * such that c[i] contains whether offset+2*i is prime or not.
     */
    fn sieve(&self, cur: usize, c: &mut BitVec) -> u64 {
        let start = sqr(self.p[cur]) + 2;
        let end = sqr(self.p[cur+1]);

        c.clear();
        c.resize(((end-start)/2) as usize, true);
        for i in 0 ..= cur {
            let p = self.p[i] as u64;
            // First odd multiple of p in range
            let mut first = p*((start-1)/p+1);
            if first % 2 == 0 {
                first += p;
            }
            let mut idx = ((first-start)/2) as usize;
            while idx < c.len() {
                let mut bit = c.get_mut(idx).unwrap();
                *bit = false;
                idx += p as usize;
            }
        }
        start
    }

    /* Compute generating primes between 9 and self.p[endidx]^2 and fill self.p with them.*/
    fn compute(&mut self, endidx: usize) {
        let mut c: BitVec = BitVec::new();
        for cur in 0 ..= endidx {
            let start = self.sieve(cur, &mut c) as usize;
            for i in 0 .. c.len() {
                if c[i] {
                    self.p.push((start+2*i) as u32);
                }
            }
        }
    }

    /* Count primes between self.p[startidx]^2 and self.p[endidx]^2.*/
    fn count(&self, startidx: usize, endidx: usize) -> usize {
        let mut c: BitVec = BitVec::new();
        let mut result: usize = 0;

        for cur in startidx .. endidx {
            self.sieve(cur, &mut c);
            result += c.count_ones();
        }
        result
    }

    fn print(&self, startidx: usize, endidx: usize, result: usize) {
        let elapsed = self.start_time.elapsed();
        println!(
            "{:>5}.{:03}s: π({:>12}²) - π({:>12}²) = {}",
            elapsed.as_secs(), elapsed.subsec_millis(),
            self.p[endidx],
            self.p[startidx],
            result,
        );
    }
}

fn main() {
    let mut calc = Calc::new();
    let mut cur: usize = 6540;
    calc.compute(cur);
    calc.print(0, cur, calc.p.len());

    let (sender, receiver) = bounded(0);

    scope(|s| {
        for _ in 0..8 {
            s.spawn(|_| {
                let receiver = receiver.clone();
                for (cur,nxt) in receiver {
                    let count = calc.count(cur, nxt);
                    calc.print(cur, nxt, count);
                }
            });
        }

        loop {
            let tgt = cur+1024;
            sender.send((cur,tgt)).unwrap();
            cur = tgt;
        }
    }).unwrap();
}
